From ce7487437bd42309014acaf8034cafe0a1e8d350 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:35:11 -0400
Subject: [block] add kblockd_schedule_delayed_work_on()

Message-id: <1402670238-13916-24-git-send-email-snitzer@redhat.com>
Patchwork-id: 83780
O-Subject: [RHEL7.1 PATCH 023/150] block: add kblockd_schedule_delayed_work_on()
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 8ab14595b6dffecea264dcca2d6d9eea7c59273a
Author: Jens Axboe <axboe@fb.com>
Date:   Tue Apr 8 09:17:40 2014 -0600

    block: add kblockd_schedule_delayed_work_on()

    Same function as kblockd_schedule_delayed_work(), but allow the
    caller to pass in a CPU that the work should be executed on. This
    just directly extends and maps into the workqueue API, and will
    be used to make the blk-mq mappings more strict.

    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-core.c b/block/blk-core.c
index fe50be2..9661c18 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -2909,6 +2909,13 @@ int kblockd_schedule_delayed_work(struct delayed_work *dwork,
 }
 EXPORT_SYMBOL(kblockd_schedule_delayed_work);
 
+int kblockd_schedule_delayed_work_on(int cpu, struct delayed_work *dwork,
+         unsigned long delay)
+{
+ return queue_delayed_work_on(cpu, kblockd_workqueue, dwork, delay);
+}
+EXPORT_SYMBOL(kblockd_schedule_delayed_work_on);
+
 #define PLUG_MAGIC 0x91827364
 
 /**
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index e66be4d..9e537d9 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -1389,6 +1389,7 @@ static inline void put_dev_sector(Sector p)
 struct work_struct;
 int kblockd_schedule_work(struct work_struct *work);
 int kblockd_schedule_delayed_work(struct delayed_work *dwork, unsigned long delay);
+int kblockd_schedule_delayed_work_on(int cpu, struct delayed_work *dwork, unsigned long delay);
 
 #ifdef CONFIG_BLK_CGROUP
 /*
-- 
1.7.1